lib/upgrade/etcd/conversionv1v3/policy.go

```diff
@@ -0,0 +1,142 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conversionv1v3
+
+import (
+	"fmt"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	"github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+)
+
+// Policy implements the Converter interface.
+type Policy struct{}
+
+// APIV1ToBackendV1 converts v1 Policy API to v1 Policy KVPair.
+func (_ Policy) APIV1ToBackendV1(a unversioned.Resource) (*model.KVPair, error) {
+	ap := a.(apiv1.Policy)
+
+	d := model.KVPair{
+		Key: model.PolicyKey{
+			Name: ap.Metadata.Name,
+		},
+		Value: &model.Policy{
+			Order:          ap.Spec.Order,
+			InboundRules:   rulesAPIV1ToBackend(ap.Spec.IngressRules),
+			OutboundRules:  rulesAPIV1ToBackend(ap.Spec.EgressRules),
+			Selector:       ap.Spec.Selector,
+			DoNotTrack:     ap.Spec.DoNotTrack,
+			Annotations:    ap.Metadata.Annotations,
+			PreDNAT:        ap.Spec.PreDNAT,
+			ApplyOnForward: ap.Spec.ApplyOnForward,
+			Types:          nil, // filled in below
+		},
+	}
+
+	if len(ap.Spec.Types) == 0 {
+		// Default the Types field according to what inbound and outbound rules are present
+		// in the policy.
+		if len(ap.Spec.EgressRules) == 0 {
+			// Policy has no egress rules, so apply this policy to ingress only.  (Note:
+			// intentionally including the case where the policy also has no ingress
+			// rules.)
+			d.Value.(*model.Policy).Types = []string{string(apiv1.PolicyTypeIngress)}
+		} else if len(ap.Spec.IngressRules) == 0 {
+			// Policy has egress rules but no ingress rules, so apply this policy to
+			// egress only.
+			d.Value.(*model.Policy).Types = []string{string(apiv1.PolicyTypeEgress)}
+		} else {
+			// Policy has both ingress and egress rules, so apply this policy to both
+			// ingress and egress.
+			d.Value.(*model.Policy).Types = []string{string(apiv1.PolicyTypeIngress), string(apiv1.PolicyTypeEgress)}
+		}
+	} else {
+		// Convert from the API-specified Types.
+		d.Value.(*model.Policy).Types = make([]string, len(ap.Spec.Types))
+		for i, t := range ap.Spec.Types {
+			d.Value.(*model.Policy).Types[i] = string(t)
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"APIv1":  ap,
+		"KVPair": d,
+	}).Debugf("Converted Policy '%s' V1 API to V1 backend", ap.Metadata.Name)
+
+	return &d, nil
+}
+
+// BackendV1ToAPIV3 converts v1 Policy KVPair to v3 API.
+func (_ Policy) BackendV1ToAPIV3(kvp *model.KVPair) (Resource, error) {
+	bp, ok := kvp.Value.(*model.Policy)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid Policy resource")
+	}
+	bk, ok := kvp.Key.(model.PolicyKey)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid Policy resource key")
+	}
+
```

Comment:
I believe the plan is that we are NOT going to convert any `knp.` policies, instead relying on the new controller to do that.

@gunjan5 can you add a check here that returns an error if a `knp.` prefixed policy is provided?

Reply:
Ok, so the wrapper code will convert them but throw them away?

I guess it's fine - just an extra precaution.  I'm fine to leave as-is.